@here/harp-mapview/lib/geometry/SolidLineMesh.ts

```diff
@@ -0,0 +1,451 @@
+}
+
+/**
+ * Finds the intersection of a ray with a extruded line.
+ * @param ray Intersection ray in object local space.
+ * @param line The centerline.
+ * @param vExtrusion Line extrusion vector.
+ * @param normal Extrusion plane normal.
+ * @param hWidth Extrusion half width.
+ * @returns Distance of the extruded line intersection to the ray origin.
+ */
+function intersectExtrudedLine(
+    ray: THREE.Ray,
+    line: THREE.Line3,
+    vExtrusion: THREE.Vector3,
+    normal: THREE.Vector3,
+    hWidth: number
+): number {
+    const matrix = new THREE.Matrix4().makeBasis(
+        line.delta(new THREE.Vector3()).normalize(),
+        vExtrusion,
+        normal
+    );
+    const obb = new OrientedBox3(
+        line.getCenter(tmpV4),
+        matrix,
+        new THREE.Vector3(line.distance() / 2, hWidth, hWidth)
+    );
+    if (obb.contains(ray.origin)) {
+        return 0;
+    }
+    return obb.intersectsRay(ray) ?? Infinity;
+}
+
+/**
+ * Finds the intersection of a ray with the closest end cap of a extruded line.
+ * @param ray Intersection ray in object local space.
+ * @param line The centerline.
+ * @param hWidth Extrusion half width.
+ * @returns Distance of the end cap intersection to the ray origin.
+ */
+function intersectClosestEndCap(ray: THREE.Ray, line: THREE.Line3, hWidth: number): number {
+    const sphere = new THREE.Sphere(line.start, hWidth);
+    const startCapT = sphere.containsPoint(ray.origin)
+        ? 0
+        : ray.intersectSphere(sphere, tmpV4)
+        ? tmpV4.sub(ray.origin).length()
+        : Infinity;
+    sphere.center.copy(line.end);
+    const endCapT = sphere.containsPoint(ray.origin)
+        ? 0
+        : ray.intersectSphere(sphere, tmpV4)
+        ? tmpV4.sub(ray.origin).length()
+        : Infinity;
+    return Math.min(startCapT, endCapT);
+}
+
+/**
+ * Intersects line
+ * @param ray Intersection ray in object local space.
+ * @param line The line to intersect.
+ * @param vExtrusion Line extrusion vector.
+ * @param hWidth The line's extrusion half width.
+ * @param hWidthSq The line's extrusion half width squared.
+ * @param plane The extrusion plane.
+ * @param interPlane The intersection of the ray with the extrusion plane.
+ * @param interLine The ray intersetion with the extruded line.
+ * @returns true if ray intersects the extruded line, false otherwise.
+ */
+function intersectLine(
+    ray: THREE.Ray,
+    line: THREE.Line3,
+    vExtrusion: THREE.Vector3,
+    hWidth: number,
+    hWidthSq: number,
+    plane: THREE.Plane,
+    interPlane: THREE.Vector3,
+    interLine: THREE.Vector3
+): boolean {
+    if (interPlane.equals(tmpRay.origin) && tmpRay.direction.dot(plane.normal) === 0) {
+        // Corner case: ray is coplanar to extruded line, find distance to extruded line sides
+        // and end caps.
+        const extrLineT = intersectExtrudedLine(ray, line, vExtrusion, plane.normal, hWidth);
+        const endCapT = intersectClosestEndCap(ray, line, hWidth);
```

Comment:
There is a commonly used geometry type for such cases - Capsule. Maybe we should introduce this type of geometry too for further purposes (as follow up, or just write a comment), intersection algorithm is quite simple though, see these:
http://www.iquilezles.org/www/articles/intersectors/intersectors.htm
https://www.shadertoy.com/view/Xt3SzX

Reply:
We don't really need to implement a capsule intersection here, with a sphere is enough, that's what I use in intersectClosestEndCap. If the ray intersects first the half of the sphere that intersects the extruded center line is fine, that means that extrLineT === endCapT.

Reply:
Yeah, of course I saw it, but it simply complicates entire logic. Whatever... if tested let's leave it like that...